opentelemetry-otlp/src/span.rs

```diff
@@ -424,13 +424,38 @@ impl SpanExporter {
     }
 
     /// Builds a new span exporter with the given configuration
+    ///
+    /// Endpoint selection:
+    /// * OTEL_EXPORTER_OTLP_TRACES_ENDPOINT, unchanged
+    /// * config.endpoint, unchanged, if it contains a path
+    /// * config.endpoint with /v1/traces appended
     #[cfg(feature = "http-proto")]
     pub fn new_http(config: ExportConfig, http_config: HttpConfig) -> Result<Self, crate::Error> {
-        let url: Uri = config
-            .endpoint
+        let endpoint_str = match std::env::var(OTEL_EXPORTER_OTLP_TRACES_ENDPOINT) {
+            Ok(val) => val,
+            Err(_) => config.endpoint,
+        };
+
+        let url: Uri = endpoint_str
             .parse()
             .map_err::<crate::Error, _>(Into::into)?;
 
+        // Preserve backwards compatability by checking if the endpoint
+        // has a path part, and using the endpoint unchanged in that case.
+        // If no path is supplied we append the standard path for traces.
+        // This means that non-standard location of the traces endpoint must
+        // be specified in full.
+        let url: Uri = match url.path() {
+            "" | "/" => {
+                let mut parts = url.into_parts();
+                parts.path_and_query = Some(http::uri::PathAndQuery::from_static("/v1/traces"));
+                parts.try_into().expect(
+                    "Should always be a valid URL, since we are constructing from a valid URL",
+                )
```

Comment:
Should we consider keep the query when motify the Uri? It's probably gonna be a noop for official OTEL collector but some users may have custom implementation that compatible with OTEL protocol

Reply:
Interesting point I hadn't thought about.  As I stated in the commit message I leave the URL alone if any path is specified for backwards compatibility with code that already has the full URL specified.  I assumed that was sufficient.

I don't have enough experience with OTEL in the wild to know if these are issues we need to address.  Maybe we should just always append /v1/traces to the path, and keep the query as-is?

I think the possibilities are:
* Keep it as is, if you need anything else than the default supplies, you must supply the full path to the trace endpoint.
* Append /v1/traces to all paths if /v1/traces is not a postfix of the path, and keep the query parts.

The second option might be an issue if a collector does not use the /v1/traces postfix.

Reply:
FWIW, both `tonic` traces implementation and metrics implementation just append the path as a string disregarding the query completely. IMHO, we should do the same here to keep the same pattern.

Reply:
Yes, I did notice that but feared someone might be relying on the current implementation, and starting to replace the path unconditionally might break existing code.  It would be great to get some input from the OTEL team on how to proceed.  I'm fine with both solutions as long as I don't need to supply the path part.

Reply:
Sorry for the late response. Since `url.path()` only output path information without any query information. I am concerned if users try to config something like `https://otel-collector.net?page=2` we will add a path for them. In this case I think we should follow the same pattern to leave it as it is.

Reply:
@TommyCpp - Just to clarify, should we set the path if neither path nor query are set? Or were you thinking of some other option?